### DISCUSSION TITLE
Recommended classes for different shades of text

### DISCUSSION THREAD

**Discussion by vnugent** [saadeghi/daisyui#2086] [Q&A] [answered]
Hi, 
I have a text heavy site and I'm using `base-content` for all my text.  I want to de-emphasize some texts, what would you recommend for defining different shades of text?  

In the Table example, I see that `opacity-50` is used to de-emphasize country.  Is it the recommended way?
https://daisyui.com/components/table/#table-with-visual-elements

Daisy semantic color classes
https://daisyui.com/docs/colors/

**Comment by saadeghi** [comment, marked as answer]
If your background is a solid color, there's no issue with using different opacities because `base-content` with opacity on top of `base-100` creates a good color mix. You can also use the shorthand classes like `base-content/20`, `base-content/50`, etc.

You can also add a new color name if you want: https://github.com/saadeghi/daisyui/discussions/653#discussioncomment-2438428

**Reply by vnugent** [reply]
Sorry to bring up this old topic again.  Using opacity such as `base-content/20` definitely works for de-emphasizing text. However, I'm looking for more of a best practice since you've worked with a lot of design systems.  One of the benefits of Daisy is that we're using less custom classes, be it our own or Tailwind's.  If we use opacity then we have to come up with new classes and have to make sure the text variants work when switching themes.

**Reply by saadeghi** [reply]
> If we use opacity then we have to come up with new classes and have to make sure the text variants work when switching themes.

@vnugent The `/20` postfix is a really useful opacity modifier from Tailwind CSS. At the end, if you want the freedom to have different shades of a color, you need a new class for that. And you don't have to worry about other themes. For example `base-content` has a good contrast with `base-100` so even decreasing the opacity to 40 gives you a readable text.

Another way is to use `@apply` to make custom class names:  
https://play.tailwindcss.com/OomOqKQLS6?file=css

Another way is to add a new class name that works with daisyUI themes:  
https://daisyui.com/blog/how-to-add-new-colors-to-daisyui/

Another way is using [CSS color-mix()](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/color-mix) to mix 2 colors together. For example `primary` and `base-100`:  
https://play.tailwindcss.com/JWOLdV6gOo  
(You can `@apply` this long class name in CSS and use a shorter custom class name)

I was planning to add color lightness modifier classes for daisyUI 4.0 but i'm not sure if Tailwind CSS itself has a plan for that or a plan for adding `color-mix()` utility classes. I don't want to make a conflict so we have to wait for the next version of Tailwind before making a decisions.

Let me know if you have any questions.

**Reply by vnugent** [reply]
> At the end, if you want the freedom to have different shades of a color, you need a new class for that. And you don't have to worry about other themes. For example base-content has a good contrast with base-100 so even decreasing the opacity to 40 gives you a readable text.

What is your recommended gradation for base-100, 200 and 300?  As you can see in my example, the helper text, text-base-300, is too light.

<img width="715" alt="Screenshot 2023-11-12 at 8 06 13 AM" src="https://github.com/saadeghi/daisyui/assets/3805254/cd126fa7-00f5-4ad9-ae29-d4d29afd4905">

**Reply by saadeghi** [reply]
@vnugent something like `text-base-content/50` would look good.
Generally `*-content` colors are better for text and you can modify the opacity with a `/`

Example: https://play.tailwindcss.com/zeHdw1Yrif

Let me know if you have any questions.

**Reply by vnugent** [reply]
https://m2.material.io/design/color/text-legibility.html#text-types (can't find it in the latest v3)

I'm going to create a custom class, `text-helper` = text-base-content/60.  I hope you'll introduce some kind of semantic text types in the future.

**Reply by saadeghi** [reply]
>I hope you'll introduce some kind of semantic text types in the future.

Sure. If the upcoming tailwind version 4.0 doesn't provide a way to dynamically change the lightness, I will. Because otherwise it would make a conflict with Tailwind CSS itself and we don't want that.

An alternative way:  

With the daisyUI 4 update and the new CSS [color-mix()](https://caniuse.com/mdn-css_types_color_color-mix) function, you can use something like this to make the base content text dim:
```
text-[color-mix(in_oklab,oklch(var(--bc))_50%,oklab(var(--b1)))]
```
The above class name gives you a new color my mixing `base-100` and `base-content` on 50%.
It's a long arbitrary class name but it works really good and everything is customizable in it:
- oklab is a more natural way to mix to colors and it gives a satisfying result
- OKLCH is now the default color model in daisyUI 4
- `--b1` is base-100 color and `--bc` is base-content. It can be any other color.
